components/Rate/src/index.svelte

```diff
@@ -0,0 +1,176 @@
+				return previous;
+			}, {});
+		}
+	};
+
+	// text or score, text will cover score
+	let text = '';
+	const textsMap = getContinuity(maxNumbers, texts);
+	$: {
+		if (showText) {
+			text = isString(texts) ? texts : textsMap[Math.ceil(currentValue)];
+		} else if (showScore) {
+			text = scoreTemplate.replace(
+				/\{\s*value\s*\}/,
+				rateDisabled ? `${value}` : `${currentValue}`
+			);
+		}
+	}
+
+	const setCurrentValue = (value: number) => {
+		if (rateDisabled) {
+			return;
+		}
+		currentValue = value;
+	};
+
+	const resetCurrentValue = () => {
+		if (rateDisabled) {
+			return;
+		}
+		currentValue = value;
+	};
+
+	$: valueDecimal = value * 100 - Math.floor(value) * 100;
+	$: valueDecimalLeftWidth = +iconSize.replace(/px/g, '') * (valueDecimal / 100);
+	$: valueDecimalRightWidth = +iconSize.replace(/px/g, '') - valueDecimalLeftWidth;
+	const showDecimalIcon = (item: number) => {
+		return rateDisabled && valueDecimal > 0 && item - 1 < value && item > value;
+	};
+
+	const iconsMap = getContinuity(maxNumbers, icons);
+	$: activeIcon = iconsMap[Math.ceil(currentValue)];
+
+	const colorsMap = getContinuity(maxNumbers, colors);
+	$: activeColor = colorsMap[Math.ceil(currentValue)];
+
+	$: decimalIcon = iconsMap[Math.ceil(value)];
+	$: decimalActiveColor = colorsMap[Math.ceil(value)];
+	$: decimalVoidColor = voidColor;
+
+	const handleUpdateValue = () => {
+		if (rateDisabled) {
+			return;
+		}
+		dispatch('updateValue', currentValue);
+	};
+
+	// class names
+	const prefixCls = getPrefixCls('rate');
+	$: baseCls = createCls(prefixCls, { [`${prefixCls}--disabled`]: rateDisabled }, cls);
+	$: itemCls = createCls(`${prefixCls}--item`, {
+		[`${prefixCls}--item__not-disabled`]: !rateDisabled
+	});
+	$: textCls = createCls(`${prefixCls}--text`);
+
+	// decimal icon => not full checked
+	$: decimalIconLeftCls = createCls(
+		`${prefixCls}--icon__decimal`,
+		`${prefixCls}--icon__decimal-left`
+	);
+	$: decimalIconRightCls = createCls(
+		`${prefixCls}--icon__decimal`,
+		`${prefixCls}--icon__decimal-right`
+	);
+</script>
+
+<div class={baseCls} {...$$restProps} {...attrs}>
+	{#each maxNumbers as item}
+		<span
+			class={itemCls}
+			on:mousemove={() => setCurrentValue(item)}
+			on:mouseleave={resetCurrentValue}
+			on:click={handleUpdateValue}
+		>
+			{#if !showDecimalIcon(item)}
+				{#if item <= currentValue}
+					<KIcon icon={activeIcon} width={iconSize} height={iconSize} color={activeColor}></KIcon>
+				{:else}
+					<KIcon
+						icon={rateDisabled ? disableVoidIcon : voidIcon}
+						width={iconSize}
+						height={iconSize}
+						color={rateDisabled ? disableVoidColor : voidColor}
+					></KIcon>
+				{/if}
+			{:else}
+				<KIcon
+					cls={decimalIconLeftCls}
+					icon={decimalIcon}
+					width={iconSize}
+					height={iconSize}
+					color={decimalActiveColor}
+					attrs={{
+						style: `width: ${valueDecimalLeftWidth}px`
+					}}
+				></KIcon>
+				<KIcon
+					cls={decimalIconRightCls}
```

Comment:
How does Left and Right manifest, can you explain? I don't see that in the example

Reply:
> How does Left and Right manifest, can you explain? I don't see that in the example

I originally planned to implement half-selection, but the icon component was not very friendly for setting half-selection, so I put the half-selection plan on hold for the time being.